Incorrect netty server compression level handling

### Expected Behavior

Setting `micronaut.server.netty.compression-level` in `application.yml` should change compression level and be set correctly. According to netty documentation:
`compressionLevel - 1 yields the fastest compression and 9 yields the best compression. 0 means no compression. The default compression level is 6.` So value should be supplied as casual integer.

### Actual Behaviour

Setting this value causes application crash during startup. It causes error: 
`Error starting Micronaut server: Error instantiating bean of type  [io.micronaut.web.router.AnnotatedFilterRouteBuilder]`
`Caused by: java.lang.StringIndexOutOfBoundsException: begin 0, end -1, length 1`


I found workaround and root cause of this behaviour.
`compressionLevel` is declared in `NettyHttpServerConfiguration` as readable bytes, which means it expects some value followed by size suffix (such as 'KB', 'GB')
Then, `readableBytesTypeConverter` tries to convert CharSequence in the following way:
```
 String value = object.toString().toUpperCase(Locale.ENGLISH);
        try {
            long numberPart = Long.parseLong(value.substring(0, value.length() - 2));
            long size;
            if (value.endsWith("KB")) {
                size = numberPart * KB_UNIT;
            } else if (value.endsWith("MB")) {
                size = numberPart * KB_UNIT * KB_UNIT;
            } else if (value.endsWith("GB")) {
                size = numberPart * KB_UNIT * KB_UNIT * KB_UNIT;
            } else {
                size = Long.parseLong(value);
            }
            return ConversionService.SHARED.convert(size, targetType);
            ...
 ```
 
 That means te `numberPart` is actually negative length and causes error later on. A simple workaround for now is to just set this value to `00x` where `x` is your desired value. 
 
 In previous micronaut versions (3.9.0) `readableBytesTypeConverter` was different and didn't cause errors:
 
![image](https://github.com/micronaut-projects/micronaut-core/assets/38864852/d35300e4-8a7c-479b-8e88-1f9851f8744f)


### Steps To Reproduce

_No response_

### Environment Information

_No response_

### Example Application

_No response_

### Version

4.2.3

--- a/core/src/main/java/io/micronaut/core/convert/format/ReadableBytesTypeConverter.java
+++ b/core/src/main/java/io/micronaut/core/convert/format/ReadableBytesTypeConverter.java
@@ -30,6 +30,9 @@ import java.util.Optional;
  */
 public class ReadableBytesTypeConverter implements FormattingTypeConverter<CharSequence, Number, ReadableBytes> {
 
+    private static final String KILOBYTES = "KB";
+    private static final String MEGABYTES = "MB";
+    private static final String GIGABYTES = "GB";
     private static final int KB_UNIT = 1024;
 
     @Override
@@ -44,13 +47,15 @@ public class ReadableBytesTypeConverter implements FormattingTypeConverter<CharS
         }
         String value = object.toString().toUpperCase(Locale.ENGLISH);
         try {
-            long numberPart = Long.parseLong(value.substring(0, value.length() - 2));
             long size;
-            if (value.endsWith("KB")) {
+            if (value.endsWith(KILOBYTES)) {
+                long numberPart = parseSizeWithUnit(value);
                 size = numberPart * KB_UNIT;
-            } else if (value.endsWith("MB")) {
+            } else if (value.endsWith(MEGABYTES)) {
+                long numberPart = parseSizeWithUnit(value);
                 size = numberPart * KB_UNIT * KB_UNIT;
-            } else if (value.endsWith("GB")) {
+            } else if (value.endsWith(GIGABYTES)) {
+                long numberPart = parseSizeWithUnit(value);
                 size = numberPart * KB_UNIT * KB_UNIT * KB_UNIT;
             } else {
                 size = Long.parseLong(value);
@@ -60,5 +65,9 @@ public class ReadableBytesTypeConverter implements FormattingTypeConverter<CharS
             context.reject(value, e);
             return Optional.empty();
         }
+    }
+
+    private long parseSizeWithUnit(String value) {
+        return Long.parseLong(value.substring(0, value.length() - 2));
     }
 }

--- a/core/src/test/groovy/io/micronaut/core/convert/format/ReadableBytesTypeConverterSpec.groovy
+++ b/core/src/test/groovy/io/micronaut/core/convert/format/ReadableBytesTypeConverterSpec.groovy
@@ -1,0 +1,28 @@
+package io.micronaut.core.convert.format
+
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class ReadableBytesTypeConverterSpec extends Specification {
+
+    @Unroll
+    void "test conversion readable bytes"() {
+        given:
+        def converter = new ReadableBytesTypeConverter()
+
+        when:
+        def converted = converter.convert(value, Long.class, null).orElse(null)
+
+        then:
+
+        converted
+        converted == mustBe
+
+        where:
+        value    | mustBe
+        "100kb"  | 100 * 1024L
+        "1000Gb" | 1000 * 1024 * 1024 * 1024L
+        "10"     | 10L
+        "10MB"   | 10 * 1024 * 1024L
+    }
+}

